src/v/utils/chunked_kv_cache.h

```diff
@@ -0,0 +1,195 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#pragma once
+
+#include "container/chunked_hash_map.h"
+#include "utils/s3_fifo.h"
+
+#include <seastar/core/shared_ptr.hh>
+#include <seastar/util/optimized_optional.hh>
+
+#include <boost/intrusive/list.hpp>
+#include <boost/intrusive/options.hpp>
+#include <boost/intrusive/parent_from_member.hpp>
+
+namespace utils {
+
+/**
+ * A basic key-value cache implementation built on top of the s3_fifo::cache.
+ */
+template<
+  typename Key,
+  typename Value,
+  typename Hash = std::conditional_t<
+    detail::has_absl_hash<Key>,
+    detail::avalanching_absl_hash<Key>,
+    ankerl::unordered_dense::hash<Key>>,
+  typename EqualTo = std::equal_to<Key>>
+class chunked_kv_cache {
+    struct cached_value;
+    struct evict;
+    using cache_t = s3_fifo::cache<
```

Comment:
nit: a note on naming: this thing you are building is the cache, and the `s3_fifo::cache` is really the eviction policy. i think i shouldn't have named it `cache` originally. may or may not be worth fixing in a follow-up PR or tacking on a rename commit to the end of this PR.

Reply:
If I end up having to iterate on other commits I'll tack a commit renaming `s3_fifo::cache` to `s3_fifo::policy` to the end of this PR. Otherwise I'll post a follow up PR.